cpp/src/parquet/encryption/key_toolkit.cc

```diff
@@ -0,0 +1,53 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/encryption/key_toolkit.h"
+
+namespace parquet {
+namespace encryption {
+
+std::shared_ptr<KmsClient> KeyToolkit::GetKmsClient(
+    const KmsConnectionConfig& kms_connection_config, uint64_t cache_entry_lifetime_ms) {
+  if (kms_client_factory_ == NULL) {
+    throw ParquetException("No KmsClientFactory is registered.");
+  }
+  std::shared_ptr<ConcurrentMap<std::shared_ptr<parquet::encryption::KmsClient>>>
+      kms_client_per_kms_instance_cache =
```

Comment:
I'm saying this once, but instead of explicitly declaring variables like this, you can (and should, when the type name is long like this) just use `auto`:
```c++
auto kms_client_per_kms_instance_cache = ... 
```